base/compiler/abstractinterpretation.jl

```diff
@@ -93,6 +93,13 @@ function abstract_call_gf_by_type(interp::AbstractInterpreter, @nospecialize(f),
                 end
             end
         else
+            csig = get_compileable_sig(method, sig, match.sparams)
+            if csig !== nothing && csig !== sig
+                # also infer the compilation signature for this method, so it's available
+                # to the compiler in case it ends up needing it (which is likely).
+                abstract_call_method(interp, method, csig, match.sparams, multiple_matches, sv)
+            end
```

Comment:
Afraid to increase the complexity of `AbstractInterpreter`, but I'd like to put this behind a configuration overload and make it specific to `NativeInterpreter` by default (similarly to `lock_mi_inference(interp)`), since this extra routine is probably unnecessary for code inspection tools like JET/Cthulhu.

Reply:
What we may likely want to consider is indicating here to the compiler that we will not use this result, e.g. by setting the `ssavalue_uses` set to empty around this call.
```suggestion
            csig = get_compileable_sig(method, sig, match.sparams)
            if csig !== nothing && csig !== sig
                # also infer the compilation signature for this method, so it's available
                # to the compiler in case it ends up needing it (which is likely).
                curruses = sv.ssavalue_uses[sv.currpc]
                sv.ssavalue_uses[sv.currpc] = BitSet()
                abstract_call_method(interp, method, csig, match.sparams, multiple_matches, sv)
                sv.ssavalue_uses[sv.currpc] = curruses
            end
```
This also triggers a correction to other related heuristics, for recursion handling, that can be disabled across this analysis.

Reply:
Good ideas; I will do both.